### DISCUSSION TITLE
Source and sink terms in Cahn-Hilliard equation

### DISCUSSION THREAD

**Discussion by mdh23rct** [idaholab/moose#25511] [Q&A Modules: Phase field]
Dear MOOSE experts,
Would you be kind to let me know the way of adding sink and source terms in the Cahn-Hilliard equation, to be specific in MOOSE!

I would also appreciate if you could provide me a step-by-step guideline to do so, if possible!

Thanks for your support and time.

Regards.

**Comment by GiudGiud** [comment]
Hello

How is the source or sink term calculated?
What does it depend on?

I would recommend looking through the phase field module for examples with such terms

Guillaume

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
I appreciate your reply.
1. The source term is probabilistic. looks like:
 
Pk(r,t) = B*Pi(r, t); where Pi is the probability of generation of species i.

2. The sink term is deterministic, depends on parameters, and is given by:

Rik(r,t) = Rr*Ck*Ci; where Rr is a rate parameter, Ck and Ci are the concentration of species k and i, respectively.

I have looked at the phase field module, but not finding examples that elaborate on the way of adding source and sink terms in Cahn-Allen equation.

Would you be kind to point me to the right direction! If there is a particular example that explains this, I would appreciate if you could mention.

Thanks and Best regards.

**Reply by GiudGiud** [reply]
I think you could write these terms as ADParsedMaterial
https://mooseframework.inl.gov/source/materials/ParsedMaterial.html
and then use this kernel to add the sink or source to the equation
https://mooseframework.inl.gov/source/kernels/ADMatCoupledForce.html

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
I appreciate your reply. I have other questions. 
1. I need to add a stochastic function as a source term to the Cahn-Hilliard equation. The stochastic function should generate random numbers scaled by a chosen parameter. Would you be kind to let me know, how I could do so in MOOSE?
2. I have tried to use ParsedElementDeletionGenerator. However, it shows an error:
A 'ParsedElementDeletionGenerator' is not a registered object.
If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

Please be aware that I have turned on all the modules (including tensor mechanics and phase field) in the Makefile. Still I am getting this problem. Is there any other issue!

Thanks,

**Reply by GiudGiud** [reply]
Hello

so 1) you can use the RandomInterface when coding in MOOSE
see all the APIs you get from this:
https://mooseframework.inl.gov/docs/doxygen/moose/classRandomInterface.html

2)  This mesh generator was added recently. you ll need to update moose to use it

Guillaume

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Thank you for your prompt reply. One question. I looked at the input parameters of ParsedElementDeletionGenerator. It allows using volume or co-ordinates of elements as a deletion criteria. However, I was planning to use the non-conserved order parameter as a element deletion criteria after each time step of a transient analysis. Is it allowed though?? My code looks like:
```
[Mesh]
    [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
    xmax = 500
    ymax = 500
    elem_type = QUAD
    []
    [delete]
    type = ParsedElementDeletionGenerator
    input = gmg
    expression = 'psi > 0.9'
    []
[]
```
Thanks.

**Reply by GiudGiud** [reply]
hello

no that s not allowed. Variables are not created when the mesh is being generated.
You ll have to work with this object instead:
https://mooseframework.inl.gov/source/userobjects/CutElementSubdomainModifier.html

and move these elements to a subdomain where no physics is active

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
As you instructed, I tried to implement the CutElementSubdomainModifier. My code looks like:
```
[Mesh]
    use_displaced_mesh = true
    [./gmg]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
    xmax = 500
    ymax = 500
    elem_type = QUAD
    [../]
[]

 [Modules]
 [./PhaseField]
       [./Nonconserved]
        [./psi]
          free_energy = F
          mobility = L
          kappa = kappa_c
          coupled_variables = c
        [../]
       [../]
      [../]
    []

    [XFEM]
      output_cut_plane = true
    []

[UserObjects]
  [./level_set_cut_uo]
    type = LevelSetCutUserObject
    level_set_var = psi
    heal_always = true
  [../]
  [./esm]
    type = CutElementSubdomainModifier
    geometric_cut_userobject = level_set_cut_uo
    apply_initial_conditions = false
  [../]
[]

[Functions]
  [./psi_func]
    type = ParsedFunction
    expression = 'psi>0.8'
  [../]
[]

#[AuxVariables]
  #[./psi]
 # [../]
#[] 

[Variables]
    [./psi]
      order = FIRST
     family = LAGRANGE
    [../]
  []

[AuxKernels]
  [./psi_function]
    type = FunctionAux
    variable = psi
    function = psi_func
  [../]
[]

[ICs]
[./psi_IC]
      type = ConstantIC
      variable = psi
      value = 0
    [../]
  []
```
However, when I run the code, it shows Variable 'psi' does not exist in this system. Next, if I try to make psi an auxiliary variable, it shows psi cannot be an auxiliary or non-linear variable at the same time. What I could do under this circumstance? Would you be kind to let me know!

The main objective is to remove elements from the calculation after each time step of a transient analysis, when the non-conserved order parameter (psi) of these elements reaches 0.8.

Thanks for your support.

**Reply by GiudGiud** [reply]
psi_func is missing parameters i think.
You need to be declairing the coupled_variables of a parsed expression.
so add
`coupled_variables = psi`

this might not be the only problem let me try to run it

**Reply by GiudGiud** [reply]
this is also invalid
```
[AuxKernels]
  [psi_function]
    type = FunctionAux
    variable = psi
    function = psi_func
  []
[]
```
because psi is a nonlinear variable rn.

- do you want to solve for psi?
- what do you need this psi_function for?

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Thank you for your reply. 
1. Yes, I need to solve the psi at each time step. For that I am using
```
[Modules]
     [./PhaseField]
       [./Conserved]
         [./c]
           free_energy = F
           mobility = M
           kappa = kappa_c
           solve_type = REVERSE_SPLIT
           coupled_variables = psi
         [../]
       [../]
      [../]
     [./PhaseField]
       [./Nonconserved]
        [./psi]
          free_energy = F
          mobility = L
          kappa = kappa_c
          coupled_variables = c
        [../]
       [../]
      [../]
    []
```
3. The psi_function I am using to choose the deletion threshold for elements. In this case 
```
[Functions]
  [./psi_func]
    type = ParsedFunction
    expression = 'psi>0.8'
  [../]
[]
```
I don't know if I could clarify properly.

Thanks.

**Reply by GiudGiud** [reply]
ok so because you need to solve for psi on every time step, psi needs to be a nonlinear variable.
Now that this is established, you cannot use psi as the `variable` for an auxkernel. auxkernels only act on auxiliary variables. 

This:
```
[Functions]
  [./psi_func]
    type = ParsedFunction
    expression = 'psi>0.8'
  [../]
[]
```
should be this:
```
[Functions]
  [./psi_func]
    type = ParsedFunction
    expression = 'psi>0.8'
    coupled_variables = 'psi'
  [../]
[]
```

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Thank you for your reply. I have turned off the Auxkernel and added coupled variables. Unfortunately, it is not working. The parsed function is not recognizing the expression 'psi>0.8'. Because the expression can be a function of time or coordinate, but not of the non-conserved order parameter. What I can do now? Is there any advice from your side to solve this problem?

Thanks.

**Reply by GiudGiud** [reply]
Oh you’re right my bad. Functions cannot depend on variables.

so the user object that modifies the mesh cannot use anything else than a function ?

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Thank you for your reply. I am not an expert in MOOSE. That's why I don't know other ways to eliminate elements based on the non-conserved order parameter. 

Actually, it should be straight forward to turn off the elements based on a solution variable. I did it in other solvers. Would you be kind to enquire, if any colleague of yours has solved such a problem earlier!!

Thanks.

**Reply by GiudGiud** [reply]
@dschwen @hugary1995 have you seen this done?

**Comment by GiudGiud** [comment]
I think this is the object to use
https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html

it works with a threshold on a variable, so you can use psi as the variable directly

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Thank you for your time. I have coded using the SubdomainModifier you have suggested. 
```
[Mesh]
  [gen]
    type = GeneratedMesh
    dim = 2
    nx = 10
    ny = 10
    xmax = 500
    ymax = 500
    elem_type = QUAD
  []
  [left]
    type = SubdomainBoundingBoxGenerator
    input = 'gen'
    block_id = 1
    bottom_left = '0 0 0'
    top_right = '500 500 0'
  []
  [file]
    type = FileMeshGenerator
    file = '../mesh/deletion.e'
  [] 
[]

[UserObjects]
   [./moving_element]
      type = CoupledVarThresholdElementSubdomainModifier
      coupled_var = 'psi'
      criterion_type = ABOVE
      threshold = 0.80
      subdomain_id = 1
      moving_boundary_name = moving_boundary
      execute_on = 'TIMESTEP_END'
    [../]
  []
```
However, when I run the file it shows the error:
Task add_mesh_generator is not registered to build MooseMesh derived objects.

Am I missing anything in the code??
Thanks.

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Please be aware that I am using the module:
```
[Modules]
  [./PhaseField]
      [./Conserved]
        [./c]
          free_energy = F
          mobility = M
          kappa = kappa_c
          solve_type = REVERSE_SPLIT
        [../]
      [../]
  [../]
[] 
```
with this element deletion procedure.

Thanks.

**Reply by mdh23rct** [reply]
Dear Dr. Guillaume,
Have solved the error. Had to change type = GeneratedMesh to GeneratedMeshGenerator.

No worries.

**Reply by jackeggemeyer** [reply]
@mdh23rct were you ever able to get sink and recombination working?

**Reply by DanoLV** [reply]
Hello @mdh23rct. I am trying to do add sink and recombination terms to a phase field problem too. Did you manage to make it work?

**Reply by GiudGiud** [reply]
@laagesen do you know of an example of sink / recombination for CH? This does not seem like a complicated problem

**Reply by laagesen** [reply]
A sink term can be added with the `BodyForce` or `MaskedBodyForce` kernel. We don't have a recombination kernel in MOOSE, because that is radiation-specific it is in Marmot.

**Reply by jackeggemeyer** [reply]
@laagesen Thank you so much! I am trying to recreate your paper: https://www.sciencedirect.com/science/article/pii/S0927025622004839#!
I want to add the recombination part to my phase field code. How difficult would it be to code my own kernel that takes care of the recombination of vacancies and interstitials? I have made some progress after watching the MOOSE dev seminar but I haven't made much progress due to my lack of C++ coding skill.

**Reply by GiudGiud** [reply]
@jackeggemeyer you could request access to marmot. No guarantees and it will take some time, but best start early.

https://ncrcaims.inl.gov/Identity/Account/Login

**Reply by jackeggemeyer** [reply]
@GiudGiud Thanks! I applied a little bit ago, and my application is in review. My advisor wants me to recreate some kernels for practice. Would remaking this kernel be that difficult? 
P.S.
I imagine you guys also can't give me much assistance due the kernel Im trying to recreate is restricted.

**Reply by laagesen** [reply]
I would say it would be a pretty straightforward task and a good first coding project. You can look at many of the kernels in the MOOSE phase-field module for examples. We can help out by giving feedback to your code and answering questions, as long as we are not copying/pasting code from other repos we should be fine.

**Reply by zahidhasan83** [reply]
Dear @jackeggemeyer ,
The recombination term depends upon the atomic fraction of species. If we talk about vacancies and interstititals, many paper assume the atomic fraction of interstitials negligible, thus the recombination is not included in the model. You can dump the vacancies at the grain boundaries using a sink term, and/or apply supersaturation of vancacies to generate voids inside grains. You can live with the source and sink terms considering model simplifications. Else, you generate your recombination kernel. If you need any help on setting up the model, please drop me a message. I will try to respond as soon as I can. Thanks.

**Reply by jackeggemeyer** [reply]
@laagesen & @zahidhasan83 
Thank you so much for the insight, I am working on my .c and .h file for recombination right now. When I run into an issue that is not syntax debugging I will make a thread and tag your accounts if that's alright!

**Comment by jackeggemeyer** [comment]
@zahidhasan83 I was unable to contact you in any other way or tag you on a new thread so apologies for continuing this here if you are still able to help. I am still working on my Phase Field Model with sink and recombination. I have added source and sink using MaskedBodyForce however I can only get them to sink away from each other as seen in the images below. 
![Screenshot from 2024-07-11 10-50-43](https://github.com/user-attachments/assets/b7c29626-1409-4ee5-8778-b75010b17ff2)
![sinkaway2](https://github.com/user-attachments/assets/b1c37c6c-29c8-4f62-b2fa-d392adcc7088)
![sinkaway3](https://github.com/user-attachments/assets/478f22e9-717c-4b46-8606-bccebca9ab63)

Do you hyave any idea on why this occurs? I have swapped the signs, values and many other constants but I can only get sink to work in the opposite. Any suggestions to fix this? My input file that creates this is provided below

Also once I add recombination I can validate that recombination "works" from this paper https://www.sciencedirect.com/science/article/pii/S0927025622004839. However outside of the desnity of vacaccnies converging to a value of 0.01 which it should, the behavior of the system is poor. Convergnace at the begining of the simution also becomes really bad and gets better later in the simulation. This often creating cubes which really confuse me, or diffusing into the interface creating just one phase than randomly spindollaly decomposes multiple time steps later. It also likes to decompose into a mystery -1 phase which i cant control and happens very randomly. I have added my .c and .h files if that helps. 

I understand this is a really loaded question but I am lost messing around with constants trying to get the correct behevior to happen. Do you guys have any suggestions that keep that phases more stable so they can sink arond the lattice instead of diffusing?

[sinkandrecomEggemeyer.txt](https://github.com/user-attachments/files/16236702/sinkandrecomEggemeyer.txt)
[GrandPotentialRecombinationHFILE.txt](https://github.com/user-attachments/files/16236708/GrandPotentialRecombinationHFILE.txt)
[GrandPotentialRecombinationCFILE.txt](https://github.com/user-attachments/files/16236716/GrandPotentialRecombinationCFILE.txt)

**Reply by zahidhasan83** [reply]
@jackeggemeyer Dear Jack, please have my personal e-mail address: md.zahidhasan83@yahoo.de 
I will break down the problem in small steps first. Before approaching the cluster problem, I would address a single void.
1. use the source term of vacancies and let the vacancies diffuse to the void. If the void grows in size, good.
2. next, do the opposite. if the void shrinks, its great.
3. Finally, introduce the recombination term in the model. I mean interstitial and vacancy recombination along with their source terms. Now see what happens.

The images you have attached look like redistribution of vacancies within the voids if I am correct. Larger voids are gowing at the expense of the small one. They are not all growing alltogether due to vacancy supersaturation.

For convergence, use automatic_scaling. Cube maybe related to the mesh resolution. I once had such a problem, mesh refinement solved the issue.

I will have a look at the files you have forwarded. Will come back as soon as I can.

Cheers,
Zahid
